pkg/controller/jobs/jobset/jobset_controller.go

```diff
@@ -0,0 +1,223 @@
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	jobsetapi "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	"sigs.k8s.io/kueue/pkg/util/maps"
+)
+
+var (
+	gvk           = jobsetapi.GroupVersion.WithKind("JobSet")
+	FrameworkName = "jobset.x-k8s.io/jobset"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:           SetupIndexes,
+		NewReconciler:          NewReconciler,
+		SetupWebhook:           SetupJobSetWebhook,
+		JobType:                &jobsetapi.JobSet{},
+		AddToScheme:            jobsetapi.AddToScheme,
+		IsManagingObjectsOwner: isJobSet,
+	}))
+}
+
+//+kubebuilder:rbac:groups=scheduling.k8s.io,resources=priorityclasses,verbs=list;get;watch
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update;patch
+//+kubebuilder:rbac:groups=jobset.x-k8s.io,resources=jobsets,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=jobset.x-k8s.io,resources=jobsets/status,verbs=get;update
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/finalizers,verbs=update
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch
+
+var NewReconciler = jobframework.NewGenericReconciler(func() jobframework.GenericJob { return &JobSet{} }, nil)
+
+func isJobSet(owner *metav1.OwnerReference) bool {
+	return owner.Kind == "JobSet" && strings.HasPrefix(owner.APIVersion, "jobset.x-k8s.io/v1")
+}
+
+type JobSet jobsetapi.JobSet
+
+var _ jobframework.GenericJob = (*JobSet)(nil)
+
+func fromObject(obj runtime.Object) *JobSet {
+	return (*JobSet)(obj.(*jobsetapi.JobSet))
+}
+
+func (j *JobSet) Object() client.Object {
+	return (*jobsetapi.JobSet)(j)
+}
+
+func (j *JobSet) IsSuspended() bool {
+	return pointer.BoolDeref(j.Spec.Suspend, false)
+}
+
+func (j *JobSet) IsActive() bool {
+	for i := range j.Status.ReplicatedJobsStatus {
+		if j.Status.ReplicatedJobsStatus[i].Active > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (j *JobSet) Suspend() {
+	j.Spec.Suspend = pointer.Bool(true)
+}
+
+func (j *JobSet) GetGVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (j *JobSet) PodSets() []kueue.PodSet {
+	podSets := make([]kueue.PodSet, len(j.Spec.ReplicatedJobs))
+	for index, replicatedJob := range j.Spec.ReplicatedJobs {
+		podSets[index] = kueue.PodSet{
+			Name:     replicatedJob.Name,
+			Template: *replicatedJob.Template.Spec.Template.DeepCopy(),
+			Count:    podsCount(&replicatedJob),
+		}
+	}
+	return podSets
+}
+
+func (j *JobSet) RunWithPodSetsInfo(podSetInfos []jobframework.PodSetInfo) {
+	j.Spec.Suspend = pointer.Bool(false)
+	if len(podSetInfos) != len(j.Spec.ReplicatedJobs) {
```

Comment:
I guess this shouldn't be reached given that we would invalidate the Workload.

Add a comment.

Reply:
yes and no, the info used here is extracted from the workload's admission  not podsets. done